File: FaceBook/LowestCommonAncestor.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) {
            return null;
        }
        boolean foundSelf = root.val == p.val || root.val == q.val;
        TreeNode right = lowestCommonAncestor(root.right, p, q);
        boolean foundRight = right != null;
        TreeNode left = lowestCommonAncestor(root.left, p, q);
        boolean foundLeft = left != null;

        if (foundLeft && foundRight || ((foundLeft && foundSelf) || (foundRight && foundSelf))) {
            return root;
        } else if (foundSelf) {
            return root;
        } else if (foundRight) {
            return right;
        } else if (foundLeft) {
            return left;
        } else {
            return null;
        }
    }
}

// There is a simpler solution if we assume that the other node will always be
// in the tree

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null)
            return null;
        if (root.val == p.val || root.val == q.val)
            return root;

        TreeNode l = lcaR(root.left, p, q);
        TreeNode r = lcaR(root.right, p, q);
        if (l != null && r != null)
            return root;
        if (l != null)
            return l;
        return r;
    }
}

/**
 * The key to understand here is that we want to treat any tree problem by
 * looking at a single node. So we basically have some base cases, then
 * recursive cases, and then a return case.
 * 
 * Case 1, find 2, 3: 2 / \ 1 3
 * 
 * if (cNode == null) return null; if (cNode.val == p.val || cNode.val == q.val)
 * selfNode = true bool foundRight = traverse(root.right, p, q) != null bool
 * foundLeft = traverse(root.left, p, q) != null if ((foundLeft && foundRight)
 * || (foundLeft && self || foundRight && self)) return root; // this will be
 * the final answer
 */

File: FaceBook/lowestCommonAncestor.scala
/**
  * Definition for a binary tree node.
  * class TreeNode(var _value: Int) {
  *   var value: Int = _value
  *   var left: TreeNode = null
  *   var right: TreeNode = null
  * }
  */

object Solution {
  def lowestCommonAncestor(
      root: TreeNode,
      p: TreeNode,
      q: TreeNode
  ): TreeNode = {
    traverse(root, p, q)._1
  }

  def traverse(
      cNode: TreeNode,
      p: TreeNode,
      q: TreeNode
  ): (TreeNode, Boolean) = {
    if (cNode == null) return (null, false)

    val iAmNode = p.value == cNode.value || q.value == cNode.value

    val (lNode, nodeFoundLeft) = traverse(cNode.left, p, q)
    val (rNode, nodeFoundRight) = traverse(cNode.right, p, q)

    if (lNode != null) (lNode, true)
    else if (rNode != null) (rNode, true)
    else if (nodeFoundRight && nodeFoundLeft) (cNode, true)
    else if ((nodeFoundRight || nodeFoundLeft) && iAmNode) (cNode, true)
    else (null, nodeFoundRight || nodeFoundLeft || iAmNode)
  }
}

/*
G: root: TreeNode, p: TreeNode, q: TreeNode
O: lowestCommonAncestor: TreeNode
T: O(N)
S: O(N)

Notes:
  - p will not equal q
  - p and q exist in a tree

Ex:

 */
